### PR TITLE
CORE-1250, 1251, 1253 - Team Join Request workflow

### PR DESCRIPTION
**Pull request by ash3rz**
This adds the dialogs for requesting to join a team, an admin approving or denying a team, and viewing the admin's denial response message.

This also makes those respective team notification messages clickable.  I refactored so that `NotificationMenu` uses the same `Message` component that the notification listing was using.  I ended up reusing the current notification context to hold those dialogs so they can be opened from either the menu or the listing.

Request to Join:
image
We don't have to go with that icon for `Join`, but I thought it was cute 😄 

Admin's Join Team Request review dialog:
image
image

Denial sent to user:
image

### CODE REVIEW

File: src/components/notifications/Message.js [hunk 1]
```diff
@@ -32,13 +41,22 @@ function MessageLink(props) {
...
 
     const message = getDisplayMessage(notification);
     const action = notification.payload?.action;
 
     const isJobStatusChange = action === "job_status_change";
     const isShare = action === "share";
     if (isShare || isJobStatusChange) {
+        if (notification.payload.access_url) {
+            return (
+                <ExternalLink href={props.notification.payload.access_url}>
```
Comment (ash3rz) [inline src/components/notifications/Message.js:54]: This was only in `NotiifcationMenu` for some reason, so I moved it here

File: stories/notifications/NotificationMocks.js [hunk 2]
```diff
@@ -1,7 +1,54 @@
...
+        id: "c2547d9a-6ca4-11eb-9f56-62d47aced14b",
+        text: "Team join request denied",
+        timestamp: "1613074045560",
+    },
+    payload: {
+        action: "team_join_denial",
+        admin_message: "Superheroes only!",
+        email_address: "ipcdev@cyverse.org",
+        team_name: "batman:heroes",
+    },
```
Comment (slr71) [inline stories/notifications/NotificationMocks.js:15]: 🤣 :batman-approves:

**Review by slr71 — COMMENTED** [review]
I may have caught one thing. Everything else looks good to me, but I don't do much work on the UI 😄

File: src/components/notifications/utils.js [hunk 3]
```diff
@@ -1,11 +1,29 @@
...
-        ? notification.subject
-        : notification.message.text;
+    switch (notification.type.toLowerCase()) {
+        case NotificationCategory.DATA.toLowerCase():
+            return notification.subject;
+
+        case NotificationCategory.ANALYSIS.toLowerCase():
+            return notification.email_template !== ANALYSIS_EMAIL_TEMPLATE
+                ? notification.subject
+                : notification.message.text;
```
[the remark this answers is not in the document]
Comment (ash3rz) [reply src/components/notifications/utils.js:17]: Oops, you are correct!  I'll refactor - thanks for catching!

File: src/components/teams/dialogs/JoinTeamDialog.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                                id={build(baseId, ids.BUTTONS.JOIN_BTN)}
+                                color="primary"
+                                type="submit"
+                                onClick={handleSubmit}
+                            >
+                                {t("join")}
+                            </Button>
+                        </DialogActions>
+                    </Dialog>
+                );
```
Comment (sri2k1us) [inline src/components/teams/dialogs/JoinTeamDialog.js:150]: Think we can use `DEDialog` here just to be consistent.

File: src/components/teams/form/Toolbar.js [hunk 5]
```diff
@@ -79,6 +83,20 @@ function EditTeamToolbar(props) {
                         {t("leave")}
                     </Button>
                 )}
+                {!isAdmin && !isMember && (
```
Comment (sri2k1us) [inline src/components/teams/form/Toolbar.js:86]: This boolean can be extracted into a `const` and then reused below. 
`const notInTeam = !isAdmin && !isMember`;

File: src/components/teams/form/Toolbar.js [hunk 6]
```diff
@@ -79,6 +83,20 @@ function EditTeamToolbar(props) {
...
+                        onClick={() => {
+                            setJoinTeamDlgOpen(true);
+                        }}
+                    >
+                        {t("join")}
+                    </Button>
+                )}
                 {isAdmin && !isCreatingTeam && (
                     <Button
                         classes={{ root: classes.deleteBtn }}
```
Comment (sri2k1us) [inline src/components/teams/form/Toolbar.js:102]: Same as above. Boolean check can be extracted into a `const` and re-used below.

**Review by sri2k1us — APPROVED** [review]
LGTM. I had some minor comments.

File: src/components/teams/dialogs/JoinTeamDialog.js (hunk 4 above)
Comment (ash3rz) [reply src/components/teams/dialogs/JoinTeamDialog.js:150]: Oops, not sure why I didn't use `DEDialog` in the first place... must've been on autopilot 😊 Thanks for catching!

**Comment by ash3rz** [comment]
Thanks for the reviews!

**Review by psarando — COMMENTED** [review]
LGTM 👍

File: src/components/teams/dialogs/JoinTeamDialog.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                                    color="primary"
+                                    type="submit"
+                                    onClick={handleSubmit}
+                                >
+                                    {t("join")}
+                                </Button>
+                            </>
+                        }
+                    >
+                        <Typography classes={{ root: classes.bottomPadding }}>
```
[the remark this answers is not in the document]
Comment (ash3rz) [reply src/components/teams/dialogs/JoinTeamDialog.js:89]: Ah yes, I keep forgetting about `gutterBottom` 😬 
Thanks for the review!

**Pull request merged**